File: tagalong/commands.py
"""Parse, discover, and dispatch typed slash commands.

The router is the session's command catalog: registration is the only place a
handler is attached to a name. Filtering for the live palette is pure so the
TUI never learns what a query *means* — it only renders what :func:`match_commands`
returns and submits the chosen ``/name`` back through :meth:`CommandRouter.handle`.
"""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass


@dataclass(frozen=True)
class Command:
    """One slash command, split into its name and whitespace-delimited arguments."""

    name: str
    arguments: tuple[str, ...]


@dataclass(frozen=True)
class CommandSpec:
    """A discoverable slash command: what the palette shows, not how it runs."""

    name: str
    description: str
    aliases: tuple[str, ...] = ()

    def alias_suffix(self, *, detailed: bool = False) -> str:
        """Parenthetical alias list for help text, or empty when there are none."""
        if not self.aliases:
            return ""
        joined = ", ".join(f"/{alias}" for alias in self.aliases)
        if detailed:
            return f" (aliases: {joined})"
        return f" ({joined})"

    def listing_line(self) -> str:
        """One indented catalog row for ``/help`` with no argument."""
        detail = f": {self.description}" if self.description else ""
        return f"  /{self.name}{self.alias_suffix()}{detail}"

    def detail_line(self) -> str:
        """One-line description for ``/help <name>``."""
        description = self.description or "no description"
        return f"/{self.name}{self.alias_suffix(detailed=True)}: {description}"


def command_query(text: str) -> str | None:
    """Return the filter fragment when the command palette should be open.

    The menu is open only for a single-token slash query: leading ``/``, no
    newline, and no whitespace after the slash. Once the typist has started
    arguments (``/new keep``), the menu closes so free-form args are unobstructed.
    The returned string is the case-folded name fragment without the slash.
    """
    if not text.startswith("/"):
        return None
    if "\n" in text or "\r" in text:
        return None
    rest = text[1:]
    if any(character.isspace() for character in rest):
        return None
    return rest.casefold()


def preferred_index(items: Sequence[CommandSpec], prefer: str | None) -> int:
    """Index of ``prefer`` in ``items``, or ``0`` when absent or unset."""
    if prefer is None:
        return 0
    for offset, spec in enumerate(items):
        if spec.name == prefer:
            return offset
    return 0


def _is_subsequence(query: str, name: str) -> bool:
    """True when every character of ``query`` appears in order inside ``name``."""
    position = 0
    for character in name:
        if position < len(query) and character == query[position]:
            position += 1
    return position == len(query)


def _name_score(name: str, query: str) -> tuple[int, int, int] | None:
    """Rank how well ``name`` matches ``query``. Lower tuples sort first.

    Tiers:
      0 — prefix of the name (``ne`` → ``new``)
      1 — contiguous substring
      2 — subsequence (``nw`` → ``new``)
    Within a tier, earlier occurrence and shorter names win so ``new`` beats
    ``newer`` for the same prefix.
    """
    if name.startswith(query):
        return (0, 0, len(name))
    index = name.find(query)
    if index >= 0:
        return (1, index, len(name))
    if _is_subsequence(query, name):
        return (2, 0, len(name))
    return None


def _spec_score(spec: CommandSpec, query: str) -> tuple[int, int, int, int] | None:
    """Best score across the command's name and aliases, then description."""
    best: tuple[int, int, int] | None = None
    for candidate in (spec.name, *spec.aliases):
        score = _name_score(candidate.casefold(), query)
        if score is not None and (best is None or score < best):
            best = score
    if best is not None:
        return (*best, len(spec.name))
    if query in spec.description.casefold():
        return (3, 0, len(spec.name), len(spec.name))
    return None


def match_commands(specs: Sequence[CommandSpec], query: str) -> tuple[CommandSpec, ...]:
    """Filter and rank command specs for a palette query.

    An empty query lists every command in registration order. A non-empty query
    returns only matches, ordered by match quality then name length.
    """
    if not query:
        return tuple(specs)
    ranked: list[tuple[tuple[int, int, int, int], CommandSpec]] = []
    for spec in specs:
        score = _spec_score(spec, query)
        if score is not None:
            ranked.append((score, spec))
    ranked.sort(key=lambda item: item[0])
    return tuple(spec for _, spec in ranked)


def _require_token(token: str, *, kind: str) -> str:
    """Reject blank or whitespace-containing command tokens."""
    if not token or any(character.isspace() for character in token):
        raise ValueError(f"invalid command {kind}: {token!r}")
    return token


class CommandRouter:
    """Route registered slash commands and expose them for discovery."""

    def __init__(self, display) -> None:
        self.display = display
        self._handlers: dict[str, Callable[[Command], None]] = {}
        self._specs: list[CommandSpec] = []
        self._aliases: dict[str, str] = {}

    @property
    def handlers(self) -> dict[str, Callable[[Command], None]]:
        """Registered handlers by canonical name (read-only view for callers)."""
        return self._handlers

    def register(
        self,
        name: str,
        handler: Callable[[Command], None],
        *,
        description: str = "",
        aliases: tuple[str, ...] = (),
    ) -> None:
        """Make ``/name`` (and any aliases) available for this session."""
        name = _require_token(name, kind="name")
        if name in self._handlers or name in self._aliases:
            raise ValueError(f"command already registered: {name}")
        normalized: list[str] = []
        for alias in aliases:
            alias = _require_token(alias, kind="alias")
            if alias == name:
                raise ValueError(f"alias repeats command name: {alias}")
            if alias in self._handlers or alias in self._aliases:
                raise ValueError(f"command alias already registered: {alias}")
            if alias in normalized:
                raise ValueError(f"duplicate command alias: {alias}")
            normalized.append(alias)
        self._handlers[name] = handler
        self._specs.append(
            CommandSpec(name=name, description=description, aliases=tuple(normalized))
        )
        for alias in normalized:
            self._aliases[alias] = name

    def specs(self) -> tuple[CommandSpec, ...]:
        """Every registered command, in registration order."""
        return tuple(self._specs)

    def resolve(self, token: str) -> str:
        """Map a typed name or alias to the canonical command name."""
        if token in self._handlers:
            return token
        return self._aliases.get(token, token)

    def lookup(self, token: str) -> CommandSpec | None:
        """Resolve a name or alias to its :class:`CommandSpec`, if registered."""
        name = self.resolve(token)
        for spec in self._specs:
            if spec.name == name:
                return spec
        return None

    def match(self, query: str) -> tuple[CommandSpec, ...]:
        """Filter this session's catalog for the palette."""
        return match_commands(self._specs, query)

    def handle(self, text: str) -> None:
        """Parse a slash command and pass it to its handler, if one is registered."""
        words = text.removeprefix("/").split()
        raw_name = words[0] if words else ""
        name = self.resolve(raw_name)
        handler = self._handlers.get(name)
        if handler is None:
            self.display.note(f"unknown command: {text}")
            return
        handler(Command(name, tuple(words[1:])))

File: electron/renderer/commands.js
/**
 * Slash-command palette matching, ported from tagalong/commands.py.
 *
 * The session owns the catalog (`commands.list` over the socket); the client
 * owns only which rows a query shows and in what order. That ranking is
 * duplicated here rather than asked for, because a keystroke cannot afford a
 * round trip — so the rules below must stay the ones in `commands.py`:
 * `command_query`, `_name_score`, `_spec_score`, `match_commands`.
 */

/**
 * @typedef {object} CommandSpec
 * @property {string} name
 * @property {string} summary
 * @property {string[]} aliases
 * @property {string | null} action_id
 */

/**
 * The filter fragment when the palette should be open, else null.
 *
 * Open only for a single-token slash query: leading `/`, no newline, no
 * whitespace after the slash. Once arguments start (`/new keep`) the menu
 * closes so free-form args are unobstructed.
 * @param {string} text
 * @returns {string | null}
 */
export function commandQuery(text) {
  if (typeof text !== "string" || !text.startsWith("/")) {
    return null;
  }
  if (text.includes("\n") || text.includes("\r")) {
    return null;
  }
  const rest = text.slice(1);
  if (/\s/.test(rest)) {
    return null;
  }
  return rest.toLowerCase();
}

/** True when every character of `query` appears in order inside `name`. */
function isSubsequence(query, name) {
  let position = 0;
  for (const character of name) {
    if (position < query.length && character === query[position]) {
      position += 1;
    }
  }
  return position === query.length;
}

/**
 * Rank how well a name matches. Lower sorts first; null is no match.
 * Tier 0 prefix, 1 substring, 2 subsequence — then earliest, then shortest.
 * @returns {number[] | null}
 */
function nameScore(name, query) {
  if (name.startsWith(query)) {
    return [0, 0, name.length];
  }
  const index = name.indexOf(query);
  if (index >= 0) {
    return [1, index, name.length];
  }
  if (isSubsequence(query, name)) {
    return [2, 0, name.length];
  }
  return null;
}

function lower(value) {
  return String(value ?? "").toLowerCase();
}

function compare(left, right) {
  for (let i = 0; i < Math.max(left.length, right.length); i += 1) {
    const a = left[i] ?? 0;
    const b = right[i] ?? 0;
    if (a !== b) {
      return a - b;
    }
  }
  return 0;
}

/**
 * Best score across a command's name and aliases, then its description.
 * @param {CommandSpec} spec
 * @param {string} query
 * @returns {number[] | null}
 */
export function specScore(spec, query) {
  let best = null;
  for (const candidate of [spec.name, ...(spec.aliases ?? [])]) {
    const score = nameScore(lower(candidate), query);
    if (score !== null && (best === null || compare(score, best) < 0)) {
      best = score;
    }
  }
  const name = String(spec.name ?? "");
  if (best !== null) {
    return [...best, name.length];
  }
  if (query !== "" && lower(spec.summary).includes(query)) {
    return [3, 0, name.length, name.length];
  }
  return null;
}

/**
 * Filter and rank a catalog for a palette query. An empty query lists
 * everything in registration order.
 * @param {CommandSpec[]} specs
 * @param {string} query
 * @returns {CommandSpec[]}
 */
export function matchCommands(specs, query) {
  const catalog = Array.isArray(specs) ? specs : [];
  if (!query) {
    return [...catalog];
  }
  return catalog
    .map((spec, index) => ({ spec, index, score: specScore(spec, query) }))
    .filter((row) => row.score !== null)
    .sort((a, b) => compare(a.score, b.score) || a.index - b.index)
    .map((row) => row.spec);
}

/**
 * Validate one `commands.list` row. A malformed row is dropped rather than
 * rendered half-built.
 * @param {unknown} value
 * @returns {CommandSpec | null}
 */
export function parseCommandSpec(value) {
  if (value === null || typeof value !== "object") {
    return null;
  }
  const record = /** @type {Record<string, unknown>} */ (value);
  if (typeof record.name !== "string" || record.name === "") {
    return null;
  }
  return {
    name: record.name,
    summary: typeof record.summary === "string" ? record.summary : "",
    aliases: Array.isArray(record.aliases)
      ? record.aliases.filter((alias) => typeof alias === "string")
      : [],
    action_id: typeof record.action_id === "string" ? record.action_id : null,
  };
}

/**
 * Read a `commands.list` reply into specs, tolerating either the bare array
 * or the `{commands: [...]}` envelope the socket sends.
 * @param {unknown} value
 * @returns {CommandSpec[]}
 */
export function parseCommandList(value) {
  const rows =
    Array.isArray(value) ||
    value === null ||
    typeof value !== "object" ||
    !Array.isArray(/** @type {Record<string, unknown>} */ (value).commands)
      ? value
      : /** @type {Record<string, unknown>} */ (value).commands;
  if (!Array.isArray(rows)) {
    return [];
  }
  const specs = [];
  for (const row of rows) {
    const spec = parseCommandSpec(row);
    if (spec !== null) {
      specs.push(spec);
    }
  }
  return specs;
}

/** Keep a selection inside a list that just changed under it. */
export function clampIndex(index, length) {
  if (length <= 0) {
    return 0;
  }
  return ((index % length) + length) % length;
}

/**
 * Resolve a typed `/name …args` against the session catalog.
 *
 * Only the first token is the command; anything after is arguments. The
 * palette closes once arguments appear (`commandQuery`), but Enter on a
 * finished line still has to find the same entry.
 *
 * @param {CommandSpec[]} specs
 * @param {string} text
 * @returns {CommandSpec | null}
 */
export function findCommand(specs, text) {
  const trimmed = String(text ?? "").trim();
  if (!trimmed.startsWith("/")) {
    return null;
  }
  const token = trimmed.slice(1).split(/\s+/)[0]?.toLowerCase() ?? "";
  if (token === "") {
    return null;
  }
  return (
    specs.find(
      (spec) =>
        spec.name.toLowerCase() === token ||
        spec.aliases.some((alias) => alias.toLowerCase() === token),
    ) ?? null
  );
}

/**
 * Whitespace-delimited arguments after the command name.
 *
 * @param {string} text
 * @returns {string[]}
 */
export function slashArguments(text) {
  const trimmed = String(text ?? "").trim();
  if (!trimmed.startsWith("/")) {
    return [];
  }
  const parts = trimmed.slice(1).split(/\s+/).filter(Boolean);
  return parts.slice(1);
}

/**
 * One-line description for `/help <name>`, matching CommandSpec.detail_line.
 *
 * @param {CommandSpec} spec
 * @returns {string}
 */
export function detailLine(spec) {
  const description = spec.summary || "no description";
  const aliases = Array.isArray(spec.aliases) ? spec.aliases : [];
  const aliasPart =
    aliases.length === 0
      ? ""
      : ` (aliases: ${aliases.map((alias) => `/${alias}`).join(", ")})`;
  return `/${spec.name}${aliasPart}: ${description}`;
}

/**
 * @typedef {{ kind: "command", spec: CommandSpec, args: string[] }
 *   | { kind: "message", text: string }
 *   | { kind: "info", text: string }
 *   | { kind: "help" }
 *   | { kind: "error", text: string }} SubmitDecision
 */

/**
 * Decide what Enter on a compose line should do.
 *
 * Strips once here so classification and the message wire payload share one
 * value (TUI parity: `event.message.text.strip()` before both branches).
 * Leading-space slash lines become commands; trailing spaces do not ride
 * `message.send`.
 *
 * @param {string} rawText
 * @param {CommandSpec[]} catalog
 * @returns {SubmitDecision}
 */
export function decideSubmit(rawText, catalog) {
  const text = String(rawText ?? "").trim();
  if (!text.startsWith("/")) {
    return { kind: "message", text };
  }
  const specs = Array.isArray(catalog) ? catalog : [];
  // No bare-`/` case on purpose. A `/` always leaves the menu open, and Enter
  // then takes the highlighted row, so this function is not on that journey --
  // the TUI does the same (tui.py on_prompt_input_submitted), which is the
  // parity this issue is about.
  const spec = findCommand(specs, text);
  if (spec === null) {
    return { kind: "error", text: `unknown command: ${text}` };
  }
  const args = slashArguments(text);
  // No action_id means the command is local help (today: `/help` / `/?`).
  // With a topic, answer in-band. Palette-on-topic is wrong: it would arm a
  // highlighted row. Bare help lists every command, which is what the palette
  // already draws -- the TUI prints the same three fields per row.
  if (spec.action_id === null) {
    if (args.length === 0) {
      return { kind: "help" };
    }
    const token = String(args[0]).replace(/^\//, "");
    const topicSpec = findCommand(specs, `/${token}`);
    if (topicSpec === null) {
      return { kind: "error", text: `unknown command: /${token}` };
    }
    return { kind: "info", text: detailLine(topicSpec) };
  }
  if (args.length > 0) {
    return { kind: "error", text: `usage: /${spec.name}` };
  }
  return { kind: "command", spec, args };
}
